test/e2e/federated-ingress.go

```diff
@@ -182,13 +195,50 @@ var _ = framework.KubeDescribe("Federated ingresses [Feature:Federation]", func(
 	})
 })
 
+// Deletes all Ingresses in the given namespace name.
+func deleteAllIngressesOrFail(clientset *fedclientset.Clientset, nsName string) {
+	orphanDependents := false
+	err := clientset.Extensions().Ingresses(nsName).DeleteCollection(&v1.DeleteOptions{OrphanDependents: &orphanDependents}, v1.ListOptions{})
+	Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error in deleting ingresses in namespace: %s", nsName))
+}
+
 /*
    equivalent returns true if the two ingress spec are equivalent.
 */
 func equivalentIngress(federatedIngress, clusterIngress v1beta1.Ingress) bool {
 	return reflect.DeepEqual(clusterIngress.Spec, federatedIngress.Spec)
 }
 
+// verifyCascadingDeletionForIngress verifies that ingresses are deleted from
+// underlying clusters when orphan dependents is false and they are not deleted
+// when orphan dependents is true.
+func verifyCascadingDeletionForIngress(clientset *fedclientset.Clientset, clusters map[string]*cluster, orphanDependents *bool, nsName string) {
+	ingress := createIngressOrFail(clientset, nsName)
+	ingressName := ingress.Name
+	// Check subclusters if the ingress was created there.
+	By(fmt.Sprintf("Waiting for ingress %s to be created in all underlying clusters", ingressName))
+	waitForIngressShardsOrFail(nsName, ingress, clusters)
+
+	By(fmt.Sprintf("Deleting ingress %s", ingressName))
+	deleteIngressOrFail(clientset, nsName, ingressName, orphanDependents)
+
+	By(fmt.Sprintf("Verifying ingresses %s in underlying clusters", ingressName))
+	errMessages := []string{}
+	// ingress should be present in underlying clusters unless orphanDependents is false.
+	shouldExist := orphanDependents == nil || *orphanDependents == true
+	for clusterName, clusterClientset := range clusters {
+		_, err := clusterClientset.Extensions().Ingresses(nsName).Get(ingressName)
+		if shouldExist && errors.IsNotFound(err) {
```

Comment:
I am not a big fan of having lots of logic in tests and I still think it should be split into 3 separate functions. But it is fine for now, I guess.